rollup/rollupsyncservice/rollupsyncservice.go

```diff
@@ -0,0 +1,395 @@
+			}
+			withdrawRoot := withdrawtrie.ReadWTRSlot(rcfg.L2MessageQueueAddress, state)
+			chunks[i].Blocks[j-cr.StartBlockNumber] = &WrappedBlock{
+				Header:       block.Header(),
+				Transactions: txData,
+				WithdrawRoot: withdrawRoot,
+			}
+		}
+	}
+
+	// get metadata of parent batch: default to genesis batch metadata.
+	parentBatchMeta := &rawdb.FinalizedBatchMeta{}
+	if batchIndex > 0 {
+		parentBatchMeta = rawdb.ReadFinalizedBatchMeta(s.db, batchIndex-1)
+	}
+
+	return parentBatchMeta, chunks, nil
+}
+
+func (s *RollupSyncService) getChunkRanges(batchIndex uint64, vLog *types.Log) ([]*rawdb.ChunkBlockRange, error) {
+	if batchIndex == 0 {
+		return []*rawdb.ChunkBlockRange{{StartBlockNumber: 0, EndBlockNumber: 0}}, nil
+	}
+
+	tx, _, err := s.client.client.TransactionByHash(context.Background(), vLog.TxHash)
+	if err != nil {
+		return nil, fmt.Errorf("failed to get transaction, err: %w", err)
+	}
+
+	return s.decodeChunkBlockRanges(tx.Data())
+}
+
+// decodeChunkBlockRanges decodes chunks in a batch based on the commit batch transaction's calldata.
+func (s *RollupSyncService) decodeChunkBlockRanges(txData []byte) ([]*rawdb.ChunkBlockRange, error) {
+	const methodIDLength = 4
+	if len(txData) < methodIDLength {
+		return nil, fmt.Errorf("transaction data is too short")
+	}
+
+	method, err := s.scrollChainABI.MethodById(txData[:methodIDLength])
+	if err != nil {
+		return nil, fmt.Errorf("failed to get method by ID, ID: %v, err: %w", txData[:4], err)
+	}
+
+	values, err := method.Inputs.Unpack(txData[methodIDLength:])
+	if err != nil {
+		return nil, fmt.Errorf("failed to unpack transaction data using ABI: %v", err)
+	}
+
+	type commitBatchArgs struct {
+		Version                uint8
+		ParentBatchHeader      []byte
+		Chunks                 [][]byte
+		SkippedL1MessageBitmap []byte
+	}
+	var args commitBatchArgs
+	err = method.Inputs.Copy(&args, values)
+	if err != nil {
+		return nil, fmt.Errorf("failed to decode calldata into commitBatch args, err: %w", err)
+	}
+
+	if args.Version != batchHeaderVersion {
+		return nil, fmt.Errorf("unexpected batch version, expected: %d, got: %v", batchHeaderVersion, args.Version)
+	}
+
+	return DecodeChunkBlockRanges(args.Chunks)
+}
+
+func calculateFinalizedBatchMeta(parentBatchMeta *rawdb.FinalizedBatchMeta, batchHash common.Hash, chunks []*Chunk) rawdb.FinalizedBatchMeta {
+	totalL1MessagePopped := parentBatchMeta.TotalL1MessagePopped
+	for _, chunk := range chunks {
+		totalL1MessagePopped += chunk.NumL1Messages(totalL1MessagePopped)
+	}
+	return rawdb.FinalizedBatchMeta{
+		BatchHash:            batchHash,
+		TotalL1MessagePopped: totalL1MessagePopped,
+	}
+}
+
+// validateBatch verifies the consistency between l1 contract and l2 node data.
+// close the node and exit once any consistency check fails.
+func validateBatch(event *L1FinalizeBatchEvent, parentBatchMeta *rawdb.FinalizedBatchMeta, chunks []*Chunk, node *node.Node) error {
```

Comment:
`node` is not used anymore?

Reply:
yes. removed in [de88b8d](https://github.com/scroll-tech/go-ethereum/pull/515/commits/de88b8dd9814c2503f9be6457d162c8503448a7a)